File: data_structures/stack/stack.go
package datastructures

import "errors"

var ErrEmptyStack = errors.New("stack is empty")

type Stack struct {
	items []int
}

func NewStack() *Stack {
	return &Stack{
		items: []int{},
	}
}

func (s *Stack) Push(val int) {
	s.items = append(s.items, val)
}

func (s *Stack) Pop() (int, error) {
	if s.IsEmpty() {
		return 0, ErrEmptyStack
	}

	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]

	return top, nil
}

func (s *Stack) Peek() (int, error) {
	if s.IsEmpty() {
		return 0, ErrEmptyStack
	}

	return s.items[len(s.items)-1], nil
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack) Size() int {
	return len(s.items)
}

File: data_structures/stack/python/stack.py
class Stack:
    def __init__(self):
        self._items: list[int] = []

    def push(self, val: int) -> None:
        self._items.append(val)

    def pop(self) -> int:
        if self.is_empty():
            raise IndexError("stack is empty")
        return self._items.pop()

    def peek(self) -> int:
        if self.is_empty():
            raise IndexError("stack is empty")
        return self._items[-1]

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def size(self) -> int:
        return len(self._items)
